core/src/ModelArray.cpp

```diff
@@ -0,0 +1,332 @@
+    xArray.setData(x);
+    return xArray / y;
+}
+
+void ModelArray::setData(double value)
+{
+    resize();
+    m_data = value;
+}
+
+void ModelArray::setData(const double* pData)
+{
+    resize();
+    auto out = std::copy(pData, pData + m_sz.at(type), m_data.data());
+}
+
+void ModelArray::setData(const DataType& from) { setData(from.data()); }
+
+void ModelArray::setData(const ModelArray& from) { setData(from.m_data.data()); }
+
+void ModelArray::setDimensions(Type type, const Dimensions& newDims)
+{
+    size_t newSize = 1;
+    for (size_t dimLen : newDims) {
+        newSize *= dimLen;
+    }
+    m_dims.at(type).clear();
+    std::copy(newDims.begin(), newDims.end(), std::back_inserter(m_dims.at(type)));
+    m_sz.at(type) = newSize;
+}
+
+// Simple 1d indexing
+inline size_t indexr(const size_t* dims, const size_t nDims, size_t i) { return i; }
+
+// Special case for 2d indexing
+inline size_t indexr(const size_t* dims, const size_t nDims, size_t i, size_t j)
+{
+    return i * dims[1] + j;
+}
+
+// General case for up to 8d indexing
+inline size_t indexr(const size_t* dims, const size_t nDims, size_t i, size_t j, size_t k,
+    size_t l = 0, size_t m = 0, size_t n = 0, size_t p = 0, size_t q = 0)
+{
+    size_t stepLength = 1;
+    size_t out = 0;
+    switch (nDims) {
+    case 8:
+        out += stepLength * q;
+        stepLength *= dims[7];
+    case 7:
+        out += stepLength * p;
+        stepLength *= dims[6];
+    case 6:
+        out += stepLength * n;
+        stepLength *= dims[5];
+    case 5:
+        out += stepLength * m;
+        stepLength *= dims[4];
+    case 4:
+        out += stepLength * l;
+        stepLength *= dims[3];
+    case 3:
+        out += stepLength * k;
+        stepLength *= dims[2];
+    case 2:
+        out += stepLength * j;
+        stepLength *= dims[1];
+    case 1:
+    default:
+        out += stepLength * i;
+        return out;
+    }
+    return 0;
+}
+
+size_t indexr(const ModelArray::Dimensions& loc, const size_t* dims)
+{
+    return indexr(dims, loc.size(), loc[0], loc[1], loc[2], loc[3], loc[4], loc[5], loc[6], loc[7]);
+}
+
+const double& ModelArray::operator[](const Dimensions& dims) const
+{
+    switch (dims.size()) {
+    case (1):
+        return (*this)(dims[0]);
+    case (2):
+        return (*this)(dims[0], dims[1]);
+    case (3):
+        return (*this)(dims[0], dims[1], dims[2]);
+    case (4):
+        return (*this)(dims[0], dims[1], dims[2], dims[3]);
+    case (5):
+        return (*this)(dims[0], dims[1], dims[2], dims[3], dims[4]);
+    case (6):
+        return (*this)(dims[0], dims[1], dims[2], dims[3], dims[4], dims[5]);
+    case (7):
+        return (*this)(dims[0], dims[1], dims[2], dims[3], dims[4], dims[5], dims[6]);
+    default:
+        return (*this)(dims[0], dims[1], dims[2], dims[3], dims[4], dims[5], dims[6], dims[7]);
+    }
+}
+
+const double& ModelArray::operator()(size_t i, size_t j) const
```

Comment:
With the variadic template approach I think we can then replace all the operator() calls with a single call like this:

```
template <typename... Args>
const double& operator()(Args... args)
{
    //return m_data[index(dimensions().data(), dimensions().size(), args...)];
    return m_data[index(dimensions().data(), args...)];
}
```

Reply:
Need to check if it is legal to overload the operator though.

Reply:
My understanding is you can.